### PR TITLE
add  mynn surface scheme as opt_sfc=4 in NoahMP

### PR DESCRIPTION
**Pull request by HelinWei-NOAA**
# PR Checklist

- [ ] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [ ] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [ ] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

[issue 920 in ccpp-physics](https://github.com/NCAR/ccpp-physics/issues/920)

- [x] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.
  opt_sfc=4 will be used for some regional RTs. Results will be changed.

- [ ] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsiblity to keep the PR up-to-date with the develop branch of ufs-weather-model. 

## Description

This PR is to add MYNN surface scheme as opt_sfc=4 in NoahMP and recode the thermal roughness scheme as a subroutine. A bug with opt_sfc=1 was fixed. 2m T/Q from NoahMP instead of sfc_diag will be used as outputs.

The update is not aimed for UFS coupled prototype 8, but for RRFS.
### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!) 
- fixes #<issue_number>
- fixes noaa-emc/fv3atm/issues/<issue_number>

## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [X] WCOSS2.intel
- [ ] opnReqTest for newly added/changed feature
- [ ] CI

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs
- waiting on ccpp-physics [PR 921](https://github.com/NCAR/ccpp-physics/pull/921)
- waiting on fv3atm [PR 537](https://github.com/NOAA-EMC/fv3atm/pull/537)

### CODE REVIEW

**Comment by jkbk2004** [comment]
@HelinWei-NOAA I think we can start working on this PR. Can you sync up your branches to develop/main: ccpp and fv3atm?

File: tests/fv3_conf/compile_bsub.IN
Comment (DeniseWorthen) [inline tests/fv3_conf/compile_bsub.IN]: Please delete this file

File: tests/fv3_conf/fv3_bsub.IN
Comment (DeniseWorthen) [inline tests/fv3_conf/fv3_bsub.IN]: Please delete this file

File: tests/rt.sh
Comment (DeniseWorthen) [inline tests/rt.sh]: Please revert this change

**Comment by HelinWei-NOAA** [comment]
Which submodules are not up to date? Thanks.
> @HelinWei-NOAA please bring these up to date with respective authoritative repositories
> 
> * fv3 **NOT** up to date

**Comment by DeniseWorthen** [comment]
@SamuelTrahanNOAA Please check the changes in tests/parm/rrfs_conus13km_hrrr.nml.IN

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: Did you check the current rrfs_dev1 parallel before removing the warning that told you to check the current rrfs_dev parallel?

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: Disabling smoke is not an acceptable way to reduce the domains_stack_size required by smoke.

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: This option must be set to use the mynn surface scheme with the flake model.

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: If your code breaks smoke support, disabling the smoke test is not an acceptable way of fixing your code.

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: Did you check the current rrfs_dev1 parallel before making these changes?

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: Why are you adding a commented-out namelist option?

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (SamuelTrahanNOAA) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: Why are you removing this line?

**Comment by DusanJovic-NOAA** [comment]
Something is wrong with this PR. Description says it depends on fv3atm [PR 537](https://github.com/NOAA-EMC/fv3atm/pull/537), but .gitmodules does not point to a fork with that PR, it still points to NOAA-EMC

**Review by SamuelTrahanNOAA — CHANGES_REQUESTED** [review]

**Comment by SamuelTrahanNOAA** [comment]
Could @joeolson42 please review these changes?

**Comment by jkbk2004** [comment]
> Could @joeolson42 please review these changes?

@SamuelTrahanNOAA added @joeolson42 to the repo.

**Comment by HelinWei-NOAA** [comment]
At this moment I have reverted all opt_sfc changes. In the future we will redesign some RTs using opt_sfc=4.

**Comment by SamuelTrahanNOAA** [comment]
> At this moment I have reverted all opt_sfc changes. In the future we will redesign some RTs using opt_sfc=4.

You have not reverted the changes to the conus13km namelist.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA can you address the reviewers' comments above?

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (joeolson42) [reply tests/parm/rrfs_conus13km_hrrr.nml.IN]: I second this. This option must not be removed.

File: tests/parm/rrfs_conus13km_hrrr.nml.IN
Comment (joeolson42) [inline tests/parm/rrfs_conus13km_hrrr.nml.IN]: none of these iopt_* parameters are used in the RUC LSM, with is used in this RRFS configuration, so I'm not sure why they are being changed (or present).

**Review by joeolson42 — CHANGES_REQUESTED** [review]
Consistent with Sam's comments, the RRFS namelists need to be kept consistent with the current RRFS_dev settings.

**Comment by HelinWei-NOAA** [comment]
It is weird that I didn't make any of those changes Sam mentioned except for opt_sfc in my original [commit](https://github.com/HelinWei-NOAA/ufs-weather-model/commit/bb1c8a3333bee7656344e2a2fe41caa3eb543c79)

**Comment by SamuelTrahanNOAA** [comment]
> It is weird that I didn't make any of those changes Sam mentioned except for opt_sfc in my original [commit](https://github.com/HelinWei-NOAA/ufs-weather-model/commit/bb1c8a3333bee7656344e2a2fe41caa3eb543c79)

You may have gotten it from an earlier version of the develop branch, before the big GSL merge a few weeks ago.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA I tested to build on hera. An error occured while running ccpp_prebuild.py Can you check the compile failure?  /scratch1/NCEPDEV/stmp2/Jong.Kim/FV3_RT/rt_19438/compile_001/err

**Comment by HelinWei-NOAA** [comment]
It looks like the wrong version of rte-rrtmgp is used in my ccpp-physics repository. I am making it right now.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA One thing to check. Since changes in RTs are reverted, do you think this PR will still change baseline?

**Comment by HelinWei-NOAA** [comment]
> @HelinWei-NOAA One thing to check. Since changes in RTs are reverted, do you think this PR will still change baseline?

Yes for those tests with Noah MP. We have some refinements and bug fix for Noah MP.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA OK! I will submit auto-RT runs around noon time.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA can you test on your side? I still see some build issue on ccpp side. /glade/scratch/jongkim/UFS-RT-tests/rt-1224/FV3/ccpp/physics/physics/module_sf_noahmplsm.f90:4582:32:

 4582 |                        z0mo,z0h)
      |                                1
Error: Type mismatch in argument ‘icom’ at (1); passed INTEGER(4) to INTEGER(8)

**Comment by HelinWei-NOAA** [comment]
Fixed the issue. Compiling went through on my own test.
> @HelinWei-NOAA can you test on your side? I still see some build issue on ccpp side. /glade/scratch/jongkim/UFS-RT-tests/rt-1224/FV3/ccpp/physics/physics/module_sf_noahmplsm.f90:4582:32:
> 
> 4582 | z0mo,z0h) | 1 Error: Type mismatch in argument ‘icom’ at (1); passed INTEGER(4) to INTEGER(8)

**Comment by junwang-noaa** [comment]
@HelinWei-NOAA Just to check, do you plan to add a test with opt_sfc=4 in NoahMP in the future? Otherwise this feature may not be maintained as there is no test.

**Comment by HelinWei-NOAA** [comment]
> @HelinWei-NOAA Just to check, do you plan to add a test with opt_sfc=4 in NoahMP in the future? Otherwise this feature may not be maintained as there is no test.

They plan to use opt_sfc=4 for RRFS. What feature may not be maintained? @barlage Should we change one of RRFS RTs with opt_sfc=4?

**Comment by SamuelTrahanNOAA** [comment]
Can you copy the tests/tests/rap_control_debug to a new tests/tests file, and change iopt_sfc? Then add it to rt.sh and rt_gnu.sh next to rap_control_debug and you'll have a new test.

**Comment by HelinWei-NOAA** [comment]
@SamuelTrahanNOAA RUC not Noah MP lsm is used in the test rap_control_debug. Noah MP with opt_sfc=4 is planned for all future RRFS runs. That's why I changed those at the beginning.

**Comment by SamuelTrahanNOAA** [comment]
What about rap_noah_debug?

**Comment by barlage** [comment]
@HelinWei-NOAA @SamuelTrahanNOAA it looks like the four rrfs_v1* tests use noahmp and MYNN; should all four be set to use this new iopt_sfc=4?

**Comment by SamuelTrahanNOAA** [comment]
@joeolson42 Could you please respond to @barlage's question?

**Comment by joeolson42** [comment]
No, unfortunately I can not answer this question. I'm not sure which LSM is meant to be used. Perhaps the name of the SDF should reflect it.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA @SamuelTrahanNOAA @junwang-noaa What about creating issue to add additional tests or extensively turn on iopt_sfc=4? so that we can move on merging the PR now.

**Comment by SamuelTrahanNOAA** [comment]
As long as the conus13km tests aren't changed, I have no objection. Were it my PR though, I would want a regression test before it is merged. It is not, however, my PR nor my decision.

**Comment by jkbk2004** [comment]
If no objection, I will submit auto-RT runs to create new baseline this evening.

**Comment by HelinWei-NOAA** [comment]
The four rrfs_v1* tests use noahmp and MYNN. For consistence it is better to use MYNN scheme in Noah MP (opt_sfc=4) too. It is not about to decide which LSM is to be used. Noah MP is already used in these four RTS. @SamuelTrahanNOAA If you agree, I am going to change opt_sfc=4 for those 4 RTs. @barlage 

> No, unfortunately I can not answer this question. I'm not sure which LSM is meant to be used. Perhaps the name of the SDF should reflect it.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA Sounds like maintaining some tests with the opt_sfc=4 is acceptable according to the conversation, can you move on to update? so that we can start creating new baseline.

**Comment by HelinWei-NOAA** [comment]
> @HelinWei-NOAA Sounds like maintaining some tests with the opt_sfc=4 is acceptable according to the conversation, can you move on to update? so that we can start creating new baseline.

Done. Changed opt_sfc=4 in rap.nml.IN. Only affect those 4 rrfs_v1* tests. Noah MP is not used in the other tests using rap.nml.IN. So there is no impact on them.

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: orion
Compiler: intel
Job: BL
[BL] Repo location: /work/noaa/nems/emc.nemspara/autort/pr/941988484/20220616091510/ufs-weather-model
[BL] Error: Test rrfs_v1beta 031 failed in run_test failed 
[BL] Error: Test rrfs_v1nssl 032 failed in run_test failed 
[BL] Error: Test rrfs_v1nssl_nohailnoccn 033 failed in run_test failed 
[BL] Error: Test rrfs_v1beta_debug 069 failed in run_test failed 
Please make changes and add the following label back: orion-intel-BL

**Comment by BrianCurtis-NOAA** [comment]
> /work/noaa/nems/emc.nemspara/autort/pr/941988484/20220616091510/ufs-weather-model

137: [Orion-16-22:370810:0:370810] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xfff
ffffc2133a128)
 46: An error occurred in ccpp_physics_run for group physics, block    1 and thread    1 (ntX=   1):
 46: An error occured in noahmpdrv_run:  erreng =   157.449802856605       at i,j:            3       -9999
 46: net solar:         163.9410
 46: net longwave:       65.9686
 46: total sensible:  **********
 46: canopy evap:         0.0069
 46: ground evap:       105.8756
 46: transpiration:       0.8587
 46: total ground:    **********
 46: canopy heat stora    6.4913
 46: precip advected:     0.0001   -0.0002    0.0000    0.0004
 46:          precip:     0.0000
 46:    veg fraction:     0.1926
 46: energy budget problem in noahmp lsm

**Comment by HelinWei-NOAA** [comment]
> /work/noaa/nems/emc.nemspara/autort/pr/941988484/20220616091510/ufs-weather-model

@BrianCurtis-NOAA Can you open the permission? All rrfs_v1* tests failed?

**Comment by BrianCurtis-NOAA** [comment]
> > /work/noaa/nems/emc.nemspara/autort/pr/941988484/20220616091510/ufs-weather-model
> 
> @BrianCurtis-NOAA Can you open the permission? All rrfs_v1* tests failed?

can you access: /work/noaa/stmp/bcurtis/stmp/bcurtis/FV3_RT/rt_397081 ? that is where the runs are.

**Comment by HelinWei-NOAA** [comment]
> > > /work/noaa/nems/emc.nemspara/autort/pr/941988484/20220616091510/ufs-weather-model
> > 
> > 
> > @BrianCurtis-NOAA Can you open the permission? All rrfs_v1* tests failed?
> 
> can you access: /work/noaa/stmp/bcurtis/stmp/bcurtis/FV3_RT/rt_397081 ? that is where the runs are.

Yes. Thanks.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA when fixes are made, we can restart working on the PR.

**Comment by junwang-noaa** [comment]
@HelinWei-NOAA With the changes you made in rap.nml.IN, it looks to me all the rap tests will be impacted. Can you check?

**Comment by HelinWei-NOAA** [comment]
I am working on the issue.

**Comment by HelinWei-NOAA** [comment]
> @HelinWei-NOAA With the changes you made in rap.nml.IN, it looks to me all the rap tests will be impacted. Can you check?

@junwang-noaa No. None of rap tests use Noah MP. opt_sfc is a valid flag only when Noah MP is used.

**Comment by HelinWei-NOAA** [comment]
> @HelinWei-NOAA when fixes are made, we can restart working on the PR.

The issue has been fixed. My own 4 rrfs_v1* RTs went through. The cause is a code mistake made when I tried to address reviewer comments.

**Comment by HelinWei-NOAA** [comment]
@junwang-noaa If you want me to bring [this issue](https://github.com/ufs-community/ufs-weather-model/issues/1274) into this PR, I can make the change too.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA we are in the middle of merging #1294. @ChunxiZhang-NOAA can combine issue #1274 in his PR. So we might be able to work on this PR from tomorrow, I think. Will keep you posted.

**Comment by ChunxiZhang-NOAA** [comment]
@HelinWei-NOAA @jkbk2004 Great! Thanks!

**Comment by jkbk2004** [comment]
@HelinWei-NOAA we will start working on creating new BL. [issue 1274]( https://github.com/ufs-community/ufs-weather-model/issues/1274) will be combined with #1192.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA can you confirm it builds ok? I see compilation failure: CMake Error at FV3/ccpp/CMakeLists.txt:44 (message):
  An error occured while running ccpp_prebuild.py, check

**Comment by jkbk2004** [comment]
@HelinWei-NOAA ccpp in your branch might be behind.

**Comment by HelinWei-NOAA** [comment]
I am checking and testing now.

**Comment by HelinWei-NOAA** [comment]
@jkbk2004 now FV3 points to the right repository
> @HelinWei-NOAA ccpp in your branch might be behind.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA sure! I am working on again.

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: jet
Compiler: intel
Job: BL
[BL] Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/941988484/20220629171515/ufs-weather-model
Please make changes and add the following label back: jet-intel-BL

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: jet
Compiler: intel
Job: BL
[BL] Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/941988484/20220630030013/ufs-weather-model
[BL] Baseline creation and move successful
[RT] Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/941988484/20220630045456/ufs-weather-model
Please make changes and add the following label back: jet-intel-BL

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: orion
Compiler: intel
Job: BL
[BL] Repo location: /work/noaa/nems/emc.nemspara/autort/pr/941988484/20220630101510/ufs-weather-model
[BL] Error: Test hafs_global_1nest_atm 083 failed in run_test failed 
[BL] Error: Test hafs_global_multiple_4nests_atm 084 failed in run_test failed 
[BL] Error: Test hafs_regional_storm_following_1nest_atm_ocn 087 failed in run_test failed 
Please make changes and add the following label back: orion-intel-BL

**Comment by jkbk2004** [comment]
@HelinWei-NOAA Please, resolved the conversations. @DeniseWorthen @DusanJovic-NOAA @SamuelTrahanNOAA The PR is ready for final comments and approval. All RTs passed.

**Review by SamuelTrahanNOAA — APPROVED** [review]

**Comment by SamuelTrahanNOAA** [comment]
@jkbk2004 Github will not let me resolve the conversations. Perhaps only the author and people with write access can do that? I have re-reviewed and given my approval.

**Review by DeniseWorthen — APPROVED** [review]

**Comment by DeniseWorthen** [comment]
I've resolved what I can see of the conversations and approved.

**Comment by HelinWei-NOAA** [comment]
@jkbk2004 I didn't see any unresolved conversions. Just waiting for Dusan's approval.

**Comment by DusanJovic-NOAA** [comment]
FV3 submodule still points to fork's URL/branch.

**Comment by HelinWei-NOAA** [comment]
We need to merge ccpp-physics branch first, then FV3.
> FV3 submodule still points to fork's URL/branch.

**Comment by jkbk2004** [comment]
> We need to merge ccpp-physics branch first, then FV3.
> 
> > FV3 submodule still points to fork's URL/branch.

@HelinWei-NOAA can you work with @SamuelTrahanNOAA to merge in ccpp pr first?

**Comment by SamuelTrahanNOAA** [comment]
@jkbk2004 ccpp-physics is merged

**Comment by HelinWei-NOAA** [comment]
@SamuelTrahanNOAA Thanks. Now I need to update my FV3 branch and then someone can merge my FV3 PR.

**Comment by SamuelTrahanNOAA** [comment]
I cannot merge to fv3atm, but I'm sure there's someone else monitoring this who can merge.

**Comment by jkbk2004** [comment]
@HelinWei-NOAA can you make last update of fv3atm pointer and revert submodule pointer?

**Comment by HelinWei-NOAA** [comment]
> @HelinWei-NOAA can you make last update of fv3atm pointer and revert submodule pointer?

Done.

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA all ready to merge. Can you approve?

**Review by DusanJovic-NOAA — APPROVED** [review]

**Pull request merged**
